### ISSUE TITLE
Framework: unit tests execution time needs to be optimised

### ISSUE THREAD

**Issue by designsimply** [Automattic/wp-calypso#2108]
Raised by @gziolo 

We enabled Babel transpilation in unit tests on July 2 (7361-gh-calypso-pre-oss). At that time execution time of all tests were under 1 minute.

 `make test` with Babel:
> real    0m46.786s
user    0m42.828s
sys     0m5.281s

Execution time is growing every day with each new unit tests added. Soon it will exceed 2 minutes.

>  real    1m53.028s
user    1m43.316s
sys     0m12.601s

First of all this is great news, because that means we add many new tests. We definitely should keep that pace of increasing code coverage with unit tests. To make it happen we need to have fast feedback loop when running unit tests. The current state can cause people to skip git prepush hook (p1441786661000164-slack-calypso), which is first symptom that we are having issues in this area. There is another issue connected with release process. CircleCi checks are getting slow, which makes tempting to skip their result when releasing quick patch.

There was this idea of running the tests in parallel on CircleCI, but this solves only one issue. Maybe it's also time to revisit they way we execute full tests run.
I noticed several complains about Babel transpilation long time. It's highly possible that at the moment we transpile the same files over and over again. This is because tests are run in isolation as separate processes, so I would expect that every process takes care of transpilation on its own. We should investigate if we can share transpiled files somehow (Babel cache?).
In long run we could also take a bit different approach and use only one Mocha process to run all tests. This would require lots of effort to keep tests running properly, because we would need to make sure all `before()` and `after()` methods work without side effects that could cause errors in other tests files. With that approach we would need to warm up Mocha only once, and we would have only one Babel transpilation step.

I'm opening this issue to start discussion. This is very complicated topic which needs to be addressed as soon as possible.

**Comment by designsimply** [comment]
Summary of past comments: 

From @dmsnell 

> **webpack** has a way of running to watch and rebuild as you go, which I have heard is very fast though I'm not sure how that can sit in our build system.
> 
> I've also read that we should split off any modules that don't get rebuilt into a separate common chunk so that webpack doesn't need to repack them each time one of the regular files changes. If we aren't already doing this, it might be worth a try.

From @nb

> We can run tests in parallel on CircleCI, but we will need to take care of parallelizing them ourselves: https://circleci.com/docs/parallel-manual-setup, which shouldn't be too hard to set up. If somebody wants to give it a try, I’d be happy to help out with testing/reviewing/merging/fiddling with CircleCi settings.

> I've been digging through this part so find out what's actually slow, here's what I have figured:
>
> > I noticed several complains about Babel transpilation long time. It's highly possible that at the moment we transpile the same files over and over again. This is because tests are run in isolation as separate processes, so I would expect that every process takes care of transpilation on its own. We should investigate if we can share transpiled files somehow (Babel cache?).
>
> This is incorrect. It was the first thing that came to my mind and I have fiddled with the babel's core code and our cache is warm, running separate processes sequentially does not result in a cache miss. You can confirm that by putting a couple of `console.log` outputs in this piece: https://github.com/babel/babel/blob/master/packages/babel/src/api/register/node.js#L76-L102. Local path for this file is `node_modules/babel-core/lib/api/register/node.js`
>
> Another idea is whether spawning new processes so many times is resulting in a very long time, I don't think this is correct. Spawning a new process never takes more than a second ever. Also, my parallel testing can actually reduce to the total time spent in testing, so it's not just spawning them.
>
> The only thing remains is what happens during initialization. Since the tests have slowed down after adding `babel/register` transpiler, it's not about `mocha`, it's about `babel/register`'s initialization.
>
> A single call to `require('babel/register)` takes around 700-800ms which is certainly very long. The module loads lots of things which takes really long. There are plenty of features loaded but we don't actually need. I'm thinking maybe we can write our own loader that only calls requires `babel` and calls `babel.compile()` when there's a cache miss. If we get around that, I think the gains can be enormous.
>
> Another possible approach is to run a single mocha test with all the test files. This reduces to the total run time from >2mins to <10seconds. However, not all of the tests are compliant with this approach. Specifically, some tests mock dependent modules by adding a relative file to the test folder, which has higher priority than other paths in path resolution. Those tests should mock the dependent modules appropriately, with `proxyquire` or such. However, even after that, ~15% of the remaining tests fail due to their DOM dependencies and I couldn't figure out why. Either `lib/react-test-env-setup` does not fit well with this, or there are some tests which modify global objects and not clean up afterwards.
>
>
> > CircleCi checks are getting slow, which makes tempting to skip their result when releasing quick patch.
>
> I think this is a big big NO. Skipping pre-push is bad enough, although it's becoming common as test runs take really really really long time. But I think we should never skip CI tests. Github release a new feature called *protected branch*, we've partially enabled this to prevent force-pushes but maybe we should also prevent merging without CI tests pass.

From @ehg

> > Another possible approach is to run a single mocha test with all the test files.
> 
> A 'single' mocha instance would be good, it'll give us a lot more flexibility in modifying the test suite runner as a whole. Have you got a branch pushed where you've tried this? I have a few ideas of what will be causing errors.

From @umurkontaci 

> @ehg tree/try/superfast-gh-calypso-pre-oss Be advised that I've delete plenty of test that were using relative files to the test folder for mocking purposes. These were causing the test runner to fail with an exception instead of showing a failing test. But it should be good enough for playing.
> 
> The runner is in `bin/run-single-mocha-test-client`, I've also modified the `bin/run-all-tests` to fail-always to prevent merging it by mistake.
> 
> I have scratched a custom version of `babel/register`, which still uses `babel` for compilation, but we insert the `require` hooks and maintain our own cache. The performance gains is not much as I expected (~102 seconds). Lots of time is still spent on initialization of `babel` itself and mocha starting the tests.
> Another thing I've found. We have plenty of component which started to `es6` syntax but their extension is still `.js`. This doesn't result in performance hits whatsoever, but I think we should use the appropriate file extensions. (jsx or es6)

From @gziolo 

> Looks like work done on parallel testing (7005-gh-calypso-pre-oss) is at very advanced stage. @umurkontaci do you see it as a quick win for us?
>
> > > CircleCi checks are getting slow, which makes tempting to skip their result when releasing quick patch.
> > I think this is a big big NO. Skipping pre-push is bad enough, although it's becoming common as test runs take really really really long time. But I think we should never skip CI tests.
>
> I thought it's obvious we don't want add. That's why this issue was filed :D
>
> > Another thing I've found. We have plenty of component which started to es6 syntax but their extension is still .js. This doesn't result in performance hits whatsoever, but I think we should use the appropriate file extensions. (jsx or es6)
>
> They are going to release new version of JS specification each year starting from ES2015 (aka ES6). We could use `.babel` extension, We can't predict for how long we are going to use Babel, or how fast its performance can improve. So I think it's better to keep `.js` extension and find tests setup with Babel that satisfies us.
>
> > tree/try/superfast-gh-calypso-pre-oss Be advised that I've delete plenty of test that were using relative files to the test folder for mocking purposes. These were causing the test runner to fail with an exception instead of showing a failing test. But it should be good enough for playing.
>
> Yes, this is my favourite approach :+1: Sadly I mad correct assumption when I created this issue:
> > This would require lots of effort to keep tests running properly, because we would need to make sure all before() and after() methods work without side effects that could cause errors in other tests files.
>
> On the other hand this is not that bad. I'm sure we could introduce changes from #7005 (parallel testing) to satisfy requirements for current way we execute `make test`. At the same time we could come out with a way of migrating to single Mocha test run.
>
> Off the top of head I can propose creating pattern for test files extensions. This is often used in other OSS projects. We could enforce test file name to match pattern like `{tested-file-name}-test.js`. This way we could ensure we don't append directly tests mocks, config files or other files that simply don't contain tests. We also could incrementally add only those tests that play nicely with new test runner. That way we could maintain 2 ways of running tests until all tests work with new runner. Of course we can do it in many different ways. We should also think of pros and cons of single Mocha test run. Definitely we should think of adding fallback which would allow to run only given group of tests like we can do it now by executing `make test` in given subfolder. Though that can be easily satisfied by adding parameter to new runner.

**Comment by gziolo** [comment]
@designsimply I don't think I can spend enough time on this issue, this rather should be included in the long term plan of @rralian's team.

**Comment by designsimply** [comment]
@gziolo, thanks for the note. The assignment was just a remnant from `calypso-pre-oss`. :)

**Comment by gziolo** [comment]
Closing in favor of #3822.
